Real-Time Weather Monitoring System Enhancement

Mahmoud-B-Miqdad opened the pull request.

Mahmoud-B-Miqdad committed:
Add WeatherData class to represent weather information

Mahmoud-B-Miqdad committed:
Add ConfigurationManager singleton to load bot settings from JSON

Mahmoud-B-Miqdad committed:
Add Weather Bots (RainBot, SunBot, SnowBot) with Factory Pattern

Mahmoud-B-Miqdad committed:
Add Weather Data Parsers (JSON & XML) using Adapter Pattern

Mahmoud-B-Miqdad committed:
Added UserInteraction class to encapsulate user interactions

RandaZraik commented on WeatherMonitoringSystem/Domain/Config/BotConfig.cs:
```diff
@@ -0,0 +1,34 @@
+﻿
+namespace WeatherMonitoringSystem.Config
+{
+    using Newtonsoft.Json;
```
usually, using statements are placed outside the namespace

RandaZraik commented on WeatherMonitoringSystem/Domain/Config/BotConfig.cs:
```diff
@@ -0,0 +1,34 @@
+﻿
+namespace WeatherMonitoringSystem.Config
```
FYI, you can do `file-scoped namespace` where you don't need explicit curly blocks `{}`:
`namespace WeatherMonitoringSystem.Config;`

RandaZraik commented on WeatherMonitoringSystem/Domain/Config/ConfigurationManager.cs:
```diff
@@ -0,0 +1,56 @@
+        {
+            try
+            {
+                string relativePath = @"JSON_File\botConfig.json";
```
can you push this json file to you directory?

RandaZraik commented on WeatherMonitoringSystem/Domain/Config/ConfigurationManager.cs:
```diff
@@ -0,0 +1,56 @@
+                }
+
+                string json = File.ReadAllText(configPath);
+                dynamic config = JsonConvert.DeserializeObject<dynamic>(json);
```
We usually prefer strongly typed models not dynamic types. Lets try 
 'JsonConvert.DeserializeObject<BotsConfiguration>(json)', and configure a new model for the config model `BotsConfiguration` to have three bots configuration

RandaZraik commented on WeatherMonitoringSystem/Domain/Config/ConfigurationManager.cs:
```diff
@@ -0,0 +1,56 @@
+            }
+            catch (Exception ex)
+            {
+                throw new Exception();
```
We should be rethrowing the same exception like this `throw;`. Throwing a new empty exception will not have the original exception message nor the stack trace. Even doing `throw ex` is wrong as well not include the stack trace.

FYI, in you specific case, you are not doing anything with the exception, so you can just remove the try-catch block and let the exception  bubble up to the caller.

RandaZraik commented on WeatherMonitoringSystem/Domain/Config/ConfigurationManager.cs:
```diff
@@ -0,0 +1,56 @@
+        {
+            get
+            {
+                lock (_lock)
```
why did you have to use locks here? I don't see any type of multi-threaded contexts.

RandaZraik commented on WeatherMonitoringSystem/Domain/Bots/RainBot.cs:
```diff
@@ -0,0 +1,25 @@
+            _config = ConfigurationManager.Instance.RainBot;
+        }
+
+        public string Activate(double temperature, double humidity)
```
when activating the bot, it would make sense to return a bool to indicate whether the bot is activated or not, and leave the print statements to the caller. Same comment to the other bots.

RandaZraik replied:
we would also need to save a bool in this class to indicate whether the bot is activated or not

RandaZraik commented on WeatherMonitoringSystem/Domain/DataParsers/JsonWeatherDataParser.cs:
```diff
@@ -0,0 +1,22 @@
+            }
+            catch (Exception ex)
+            {
+                throw new Exception();
```
same comment about throwing an empty exception

RandaZraik commented:
FYI, the task specifically mentioned to use Observer and Strategy design patterns, lets revisit and think about how to utilize these patterns.

Mahmoud-B-Miqdad committed:
Refactored: Integrated **Observer & Factory Patterns** + Enhancements

- Implemented **Observer Pattern**, allowing  to notify bots automatically.
- Improved **Factory Pattern**:  now dynamically creates and registers bots.
- Removed manual bot activation, enhancing **scalability** and **maintainability**.
- Moved ** statements outside the ** for better organization.
- Used **compact  declaration** for cleaner code.
- **Refactored bot activation:**
  -  now returns a  instead of printing messages directly.
  - Added  property to track bot activation state.

Mahmoud-B-Miqdad committed:
Refactored  to follow **Strategy Pattern**

-  now **stores and switches** parsers dynamically.
- Introduced  to define the strategy once, then use .
- Updated  to allow **changing parsing strategy at runtime**.

Mahmoud-B-Miqdad committed:
Refactor: Remove unnecessary try-catch blocks for better exception propagation

Mahmoud-B-Miqdad committed:
Refactored Singleton Implementation: Replaced lock with Lazy<T> for Thread-Safety & Efficiency

Mahmoud-B-Miqdad committed:
Refactor ConfigurationManager: Replace dynamic with strongly-typed BotsConfiguration for better type safety and maintainability.

Mahmoud-B-Miqdad committed:
Added botConfig.json

RandaZraik commented on WeatherMonitoringSystem/Domain/Bots/RainBot.cs:
```diff
@@ -0,0 +1,30 @@
+public class RainBot : IWeatherBot
+{
+    private readonly BotConfig _config;
+    private bool _isActivated;
```
this is common between all bots, so lets move it to the interface please

RandaZraik commented on WeatherMonitoringSystem/Domain/Core/WeatherStation.cs:
```diff
@@ -0,0 +1,35 @@
+        public void SetWeatherData(WeatherData data)
+        {
+            _currentWeather = data;
+            NotifyBots();
```
why are we notifying the bots here? I noticed that you also do `bot.Update` again in the `UserInteraction`

RandaZraik commented on WeatherMonitoringSystem/Domain/Utilities/UserInteraction.cs:
```diff
@@ -0,0 +1,111 @@
+    {
+        PrintMessage("Welcome to the Real-Time Weather Monitoring System!");
+
+        WeatherDataParserFactory parserFactory = new WeatherDataParserFactory();
```
why factory? shouldn't we use here strategy pattern? :D

RandaZraik commented on WeatherMonitoringSystem/Domain/Utilities/UserInteraction.cs:
```diff
@@ -0,0 +1,111 @@
+        return GetUserInput("Enter data format (JSON/XML): ").Trim().ToLower();
+    }
+
+    private static WeatherData GetWeatherData(IWeatherDataParser parser)
```
since `WeatherDataParserFactory` should be a strategy and already have the parser, maybe this can be moved to that class , but we need to remove the print statements and keep the responsibility to the `UserInteraction`.

RandaZraik commented on WeatherMonitoringSystem/Domain/Utilities/UserInteraction.cs:
```diff
@@ -0,0 +1,111 @@
+        {
+            WeatherStation station = new WeatherStation();
+
+            List<IWeatherBot> bots = BotFactory.CreateBots(station);
```
this logic of activiating the bots should be in a different class, UserInteraction should just call a method to notify subscribers and that will do the whole job

RandaZraik commented on WeatherMonitoringSystem/Domain/Bots/RainBot.cs:
```diff
@@ -0,0 +1,30 @@
+        _config = ConfigurationManager.Instance.RainBot;
+    }
+
+    public bool Update(WeatherData data, out string message)
```
I know we said to not return the message, but still we shouldn't return it as `out` :D 
The idea is that you can have a message saved per bot, and in `UserInteraction` you can loop through the bots and print the state.

Mahmoud-B-Miqdad committed:
Refactor: Move common bot properties to IWeatherBot interface

Mahmoud-B-Miqdad committed:
Refactored bot notification to eliminate redundant updates

Modified NotifyBots() to return activated bots with messages.
Updated SetWeatherData() to provide activation data directly.
Removed duplicate bot.Update call in UserInteraction, improving efficiency.

Mahmoud-B-Miqdad committed:
Refactor: Implement Strategy Pattern explicitly with WeatherParserStrategy class

RandaZraik commented on WeatherMonitoringSystem/Domain/Utilities/UserInteraction.cs:
```diff
@@ -0,0 +1,110 @@
+            var activatedBots = station.SetWeatherData(weatherData);
+
+            PrintMessage("*****************************************");
+            foreach (var (bot, message) in activatedBots)
```
still not sure why we still need to return a message from the bot, you can access the message from the bot since it has now its config. It's not a good practice to return print messages from methods that have different responsibilities.

RandaZraik commented on WeatherMonitoringSystem/Domain/Bots/SnowBot.cs:
```diff
@@ -0,0 +1,31 @@
+        Config = ConfigurationManager.Instance.SnowBot;
+    }
+
+    public bool Update(WeatherData data, out string message)
```
This isn't really update the bot info, so maybe a better name is `Trigger` ;)

RandaZraik commented on WeatherMonitoringSystem/Domain/Bots/BotFactory.cs:
```diff
@@ -0,0 +1,25 @@
+
+public static class BotFactory
+{
+    public static List<IWeatherBot> CreateBots(WeatherStation station)
```
This method name indicates that it creates bots, but in the method we adding those the station. Maybe would be better to have just an extension class for `WeatherStation` and have a method called `AttachBots` or `AddBots`. Since this is more a logic related to the station.

Mahmoud-B-Miqdad committed:
Refactor: Move GetWeatherData to WeatherParserStrategy and separate user interaction concerns

Mahmoud-B-Miqdad committed:
Refactor: Remove out message, add internal Message property, rename Update() to Trigger()

Mahmoud-B-Miqdad committed:
Refactor: Separate bot creation and attachment logic

- Modified  to only create bots without attaching them to the station.
- Added  as an extension method to attach multiple bots to a station.
- Updated  to use the new methods for better separation of concerns and improved code clarity.

RandaZraik commented on WeatherMonitoringSystem/Domain/Utilities/UserInteraction.cs:
```diff
@@ -0,0 +1,103 @@
+                parserStrategy.SetParser(parser);
+               
+            }
+            catch (Exception ex)
```
`Exception` is too generic, it can be internal error messages that we shouldn't show to the user. We would usually catch specific exceptions (i.e. argument exception) that we know we will be thrown from the logic called, and we can leave the last cash to be for `Exception` and not show `ex.Message` to the user but instead something more generic `Something went wrong`. As generic exceptions could hold sensitive information.

If we don't know what type of exceptions the code would throw, we can catch `Exception` but instead of printing `ex.Message` we can do a custom exception `Something went wrong` and log the exception somewhere (we don't have to log anything now since we are not at that place in the training yet :D )

Lets revisit other try-catch blocks in this PR with similar behavior and fix :)

RandaZraik commented on WeatherMonitoringSystem/Domain/Bots/IWeatherBot.cs:
```diff
@@ -0,0 +1,14 @@
+public interface IWeatherBot
+{
+    bool Trigger(WeatherData data);
+    BotConfig Config { get; }  
```
the convention is to have properties before methods

RandaZraik commented on WeatherMonitoringSystem/Domain/Bots/RainBot.cs:
```diff
@@ -0,0 +1,30 @@
+    public BotConfig Config { get; }
+    public bool IsActivated { get; set; }
+    public string Message { get; private set; } = string.Empty;
+    public RainBot()
```
NIT: new line before the constructor

RandaZraik commented on WeatherMonitoringSystem/Domain/Bots/SunBot.cs:
```diff
@@ -0,0 +1,32 @@
+    public bool IsActivated { get; set; }
+    public string Message { get; private set; } = string.Empty;
+
+
```
NIT: remove extra new line

RandaZraik commented on WeatherMonitoringSystem/Domain/Core/WeatherStation.cs:
```diff
@@ -0,0 +1,43 @@
+            return NotifyBots();
+        }
+
+        private List<IWeatherBot> NotifyBots()
```
The name of this method doesn't indicate that it will only return the activated bots, it's a bit misleading, you can instead just return all bots here, or a separate method to return the bots.

RandaZraik commented on WeatherMonitoringSystem/Domain/DataParsers/WeatherParserStrategy.cs:
```diff
@@ -0,0 +1,27 @@
+        }
+        catch (Exception ex)
+        {
+            throw new Exception($"Error parsing JSON/XML: {ex.Message}");
```
lets avoid using the generic `Exception` type, since if the caller wanna catch `Exception` it won't know if this is from this method or a different method, we would better throw a specific type of exceptions , if nothing from the existing exception types make  sense, just do a new custom exception for your case (i.e. `XmlJsonParseException`).

Mahmoud-B-Miqdad committed:
Refactor error handling and code structure:
- Replaced generic exceptions with more specific ones to avoid exposing internal error messages.
- Updated try-catch blocks to handle specific exceptions and rethrow custom exceptions for better clarity.
- Refactored method names for better clarity and accuracy (e.g., renamed method to return all activated bots instead of just Notify them).
- Adjusted code formatting by adding/removing necessary new lines for consistency.
- Rearranged properties and methods to follow coding conventions.

RandaZraik commented on WeatherMonitoringSystem/Domain/Core/WeatherStation.cs:
```diff
@@ -0,0 +1,43 @@
+            return GetActivatedBots();
+        }
+
+        private List<IWeatherBot> GetActivatedBots()
```
renaming the method to Getxxx is still confusing, it hides with the method is actually doing, the method is doing two things: triggering the bot + returning active bots. 
Just separate it to two methods as suggested before, or return all bots from this method. It's important for the method name to be clear enough, it's fine to return all bots , but if returns just partial list , the caller can't tell that this method is doing that.

Mahmoud-B-Miqdad committed:
Enhance Exception Handling in Data Parsing

- Replaced generic 'Exception' with specific exceptions: 'JsonException' and 'XmlException'.
- Improved error messages to provide clearer feedback on JSON/XML parsing issues.

Mahmoud-B-Miqdad committed:
Refactor bot activation logic:
- separate activation and retrieval
- improve method names for clarity.

RandaZraik commented on WeatherMonitoringSystem/Domain/Utilities/UserInteraction.cs:
```diff
@@ -0,0 +1,140 @@
+            string format = GetDataFormat();
+
+            try
+            {
```
you can just do one big try with all code in while loop, and multiple catch blocks

Mahmoud-B-Miqdad committed:
Refactored error handling by nesting multiple try blocks inside a single try block and consolidating exception handling with multiple catch statements for improved readability and maintainability.
